File: benchmarks/fasta/python/trace_bench.py
# The Computer Language Benchmarks Game
# http://shootout.alioth.debian.org/
#
# modified by Ian Osgood
# modified again by Heinrich Acker
# modified by Justin Peel
# modified by Mariano Chouza
# modified by Ashley Hewson

import sys, bisect, array

alu = (
   'GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG'
   'GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA'
   'CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT'
   'ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA'
   'GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG'
   'AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC'
   'AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA')

iub = zip('acgtBDHKMNRSVWY', [0.27, 0.12, 0.12, 0.27] + [0.02]*11)

homosapiens = [
    ('a', 0.3029549426680),
    ('c', 0.1979883004921),
    ('g', 0.1975473066391),
    ('t', 0.3015094502008),
]

IM = 139968
INITIAL_STATE = 42

def makeCumulative(table):
    print "def makeCumulative(table):"
    P = []
    C = []
    prob = 0.
    for char, p in table:
        print "for char, p in table:"
        prob += p
        P += [prob]
        C += [char]
    return (P, C)

randomGenState = INITIAL_STATE
randomLUT = None
def makeRandomLUT():
    print "def makeRandomLUT():"
    global randomLUT
    ia = 3877; ic = 29573
    randomLUT = array.array("i", [(s * ia + ic) % IM for s in xrange(IM)])

def makeLookupTable(table):
    print "def makeLookupTable(table):"
    bb = bisect.bisect
    probs, chars = makeCumulative(table)
    imf = float(IM)
    return array.array("c", [chars[bb(probs, i / imf)] for i in xrange(IM)])

def repeatFasta(src, n):
    print "def repeatFasta(src, n):"
    width = 60
    r = len(src)
    s = src + src + src[:n % r]
    for j in xrange(n // width):
        print "for j in xrange(n // width):"
        i = j*width % r
        #print s[i:i+width]
        s[i:i+width]
    if n % width:
        print "if n % width:"
        #print s[-(n % width):]
        s[-(n % width):]

def randomFasta(table, n):
    print "def randomFasta(table, n):"
    global randomLUT, randomGenState
    width = 60
    rgs = randomGenState
    rlut = randomLUT
    
    lut = makeLookupTable(table)
    line_buffer = []
    la = line_buffer.append
    
    for i in xrange(n // width):
        print "for i in xrange(n // width):"
        for i in xrange(width):
            print "for i in xrange(width):"
            rgs = rlut[rgs]
            la(lut[rgs])
        #print ''.join(line_buffer)
        ''.join(line_buffer)
        line_buffer[:] = []
    if n % width:
        print "if n % width:"
        for i in xrange(n % width):
            print "for i in xrange(n % width):"
            rgs = rlut[rgs]
            la(lut[rgs])
        #print ''.join(line_buffer)
        ''.join(line_buffer)
    
    randomGenState = rgs

def run_iter(n):
    print "def run_iter(n):"
    global randomGenState
    randomGenState = INITIAL_STATE
    #n = int(sys.argv[1])

    makeRandomLUT()

    #print '>ONE Homo sapiens alu'
    repeatFasta(alu, n*2)

    #print '>TWO IUB ambiguity codes'
    randomFasta(iub, n*3)

    #print '>THREE Homo sapiens frequency'
    randomFasta(homosapiens, n*5)
    
#main()

File: benchmarks/fasta/ruby/trace_bench.rb
# The Computer Language Benchmarks Game
# http://shootout.alioth.debian.org/
# Contributed by Sokolov Yura
# Modified by Rick Branson
# Modified by YAGUCHI Yuya

INITIAL_STATE = 42.0
$last = INITIAL_STATE

GR_IM = 139968.0
GR_IA = 3877.0
GR_IC = 29573.0

@alu =
   "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"+
   "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"+
   "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"+
   "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"+
   "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"+
   "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"+
   "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

@iub = [
    ["a", 0.27],
    ["c", 0.12],
    ["g", 0.12],
    ["t", 0.27],

    ["B", 0.02],
    ["D", 0.02],
    ["H", 0.02],
    ["K", 0.02],
    ["M", 0.02],
    ["N", 0.02],
    ["R", 0.02],
    ["S", 0.02],
    ["V", 0.02],
    ["W", 0.02],
    ["Y", 0.02],
]

@homosapiens = [
    ["a", 0.3029549426680],
    ["c", 0.1979883004921],
    ["g", 0.1975473066391],
    ["t", 0.3015094502008],
]

def make_repeat_fasta(src, n)
    puts "def make_repeat_fasta(src, n)"
    v = nil
    width = 60
    l = src.length
    s = src * ((n / l) + 1)
    s.slice!(n, l)
    #puts (s.scan(/.{1,#{width}}/).join("\n"))
    s.scan(/.{1,#{width}}/).join("\n")
end

def make_random_fasta(table, n)
    puts "def make_random_fasta(table, n)"
    rand = nil
    width = 60
    chunk = 1 * width
    prob = 0.0
    rwidth = (1..width)
    table = table.collect{|v|
        puts "table.collect{|v|"
        prob += v[1]
        [v[0], prob]
    }

    collector = "rand = ($last = ($last * GR_IA + GR_IC) % GR_IM) / GR_IM\n"
    table.each do |va, vb|
      puts "table.each do |va, vb|"
      collector += "next #{va.inspect} if #{vb.inspect} > rand\n"
    end

    # Looks like eval can't deal with comments inside
    #eval <<-EOF
    #  (1..(n/width)).each do |i|
    #    puts rwidth.collect{#{collector}}.join
    #  end
    #  if n%width != 0
    #    puts (1..(n%width)).collect{#{collector}}.join
    #  end
    #EOF
    eval <<-EOF
      (1..(n/width)).each do |i|
        puts "(1..(n/width)).each do |i|"
        rwidth.collect{#{collector}}.join
      end
      if n%width != 0
        puts "if n%width != 0"
        (1..(n%width)).collect{#{collector}}.join
      end
    EOF
end

def run_iter(n)
    $last = INITIAL_STATE
    puts "def run_iter(n)"
    #n = (ARGV[0] or 27).to_i

    #puts ">ONE Homo sapiens alu"
    make_repeat_fasta(@alu, n*2)

    #puts ">TWO IUB ambiguity codes"
    make_random_fasta(@iub, n*3)

    #puts ">THREE Homo sapiens frequency"
    make_random_fasta(@homosapiens, n*5)
end
